File: src/lib/feeds.ts
import { invoke } from "@tauri-apps/api/core";

import type { TimelinePost } from "./api";

// Feed types
export interface FeedInfo {
  uri: string;
  cid: string;
  did: string;
  creator_did: string;
  creator_handle: string;
  creator_display_name: string | null;
  creator_avatar: string | null;
  display_name: string;
  description: string | null;
  avatar: string | null;
  like_count: number;
  is_saved: boolean;
}

export interface SuggestedFeedsResponse {
  feeds: FeedInfo[];
  cursor: string | null;
}

export interface FeedPostsResponse {
  posts: TimelinePost[];
  cursor: string | null;
}

// List types
export interface ListInfo {
  uri: string;
  cid: string;
  name: string;
  purpose: string;
  description: string | null;
  avatar: string | null;
  creator_did: string;
  creator_handle: string;
  creator_display_name: string | null;
  creator_avatar: string | null;
  member_count: number;
}

export interface ListMember {
  uri: string;
  did: string;
  handle: string;
  display_name: string | null;
  avatar: string | null;
}

export interface ActorListsResponse {
  lists: ListInfo[];
  cursor: string | null;
}

export interface ListDetailsResponse {
  list: ListInfo;
  members: ListMember[];
  cursor: string | null;
}

// Re-export from api.ts
export type { TimelinePost, TimelineResponse, ProfileResponse } from "./api";

// Feed functions
export async function getSuggestedFeeds(
  cursor?: string,
): Promise<SuggestedFeedsResponse> {
  return invoke<SuggestedFeedsResponse>("get_suggested_feeds", { cursor });
}

export async function getFeed(
  feedUri: string,
  limit?: number,
  cursor?: string,
): Promise<FeedPostsResponse> {
  return invoke<FeedPostsResponse>("get_feed", {
    request: { feed_uri: feedUri, limit, cursor },
  });
}

// List functions
export async function getActorLists(
  actor: string,
  cursor?: string,
): Promise<ActorListsResponse> {
  return invoke<ActorListsResponse>("get_actor_lists", { actor, cursor });
}

export async function getList(
  listUri: string,
  cursor?: string,
): Promise<ListDetailsResponse> {
  return invoke<ListDetailsResponse>("get_list", {
    request: { list_uri: listUri, cursor },
  });
}

// List membership check
export interface ListMembership {
  list_uri: string;
  listitem_uri: string;
}

export interface SubjectMembershipsResponse {
  memberships: ListMembership[];
}

export async function getSubjectListMemberships(
  subjectDid: string,
): Promise<SubjectMembershipsResponse> {
  return invoke<SubjectMembershipsResponse>("get_subject_list_memberships", {
    subjectDid,
  });
}

// List CRUD functions
export interface CreateListRequest {
  name: string;
  purpose: "curatelist" | "modlist";
  description?: string;
}

export interface CreateListResponse {
  uri: string;
  cid: string;
}

export async function createList(
  request: CreateListRequest,
): Promise<CreateListResponse> {
  return invoke<CreateListResponse>("create_list", { request });
}

export interface UpdateListRequest {
  list_uri: string;
  name: string;
  description?: string;
}

export async function updateList(request: UpdateListRequest): Promise<void> {
  return invoke<void>("update_list", { request });
}

export async function deleteList(listUri: string): Promise<void> {
  return invoke<void>("delete_list", { listUri });
}

// List member functions
export interface AddListMemberResponse {
  uri: string;
}

export async function addListMember(
  listUri: string,
  subjectDid: string,
): Promise<AddListMemberResponse> {
  return invoke<AddListMemberResponse>("add_list_member", {
    request: { list_uri: listUri, subject_did: subjectDid },
  });
}

export async function removeListMember(listitemUri: string): Promise<void> {
  return invoke<void>("remove_list_member", { listitemUri });
}

// List feed
export interface ListFeedPost {
  uri: string;
  cid: string;
  author_did: string;
  author_handle: string;
  author_display_name: string | null;
  author_avatar: string | null;
  is_repost: boolean;
  reposted_by_handle: string | null;
  reposted_by_display_name: string | null;
  text: string;
  created_at: string;
  reply_count: number;
  repost_count: number;
  like_count: number;
  is_liked: boolean;
  is_reposted: boolean;
  viewer_like: string | null;
  viewer_repost: string | null;
}

export interface ListFeedResponse {
  posts: ListFeedPost[];
  cursor: string | null;
}

export async function getListFeed(
  listUri: string,
  limit: number = 50,
  cursor?: string,
): Promise<ListFeedResponse> {
  return invoke<ListFeedResponse>("get_list_feed", {
    request: { list_uri: listUri, limit, cursor },
  });
}

File: src-tauri/src/commands/feeds.rs
use crate::commands::auth::AgentState;
use crate::error::AppError;
use crate::media;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, State};

#[derive(Serialize)]
pub struct FeedInfo {
    pub uri: String,
    pub cid: String,
    pub did: String,
    pub creator_did: String,
    pub creator_handle: String,
    pub creator_display_name: Option<String>,
    pub creator_avatar: Option<String>,
    pub display_name: String,
    pub description: Option<String>,
    pub avatar: Option<String>,
    pub like_count: u32,
    pub is_saved: bool,
}

#[derive(Serialize)]
pub struct SuggestedFeedsResponse {
    pub feeds: Vec<FeedInfo>,
    pub cursor: Option<String>,
}

/// Get suggested feeds for discovery
#[tauri::command]
pub async fn get_suggested_feeds(
    agent_state: State<'_, AgentState>,
    cursor: Option<String>,
) -> Result<SuggestedFeedsResponse, AppError> {
    let guard = agent_state.lock().await;
    let agent = guard.as_ref().ok_or(AppError::SessionNotFound)?;

    let response = agent
        .api
        .app
        .bsky
        .feed
        .get_suggested_feeds(
            bsky_sdk::api::app::bsky::feed::get_suggested_feeds::ParametersData {
                cursor,
                limit: None,
            }
            .into(),
        )
        .await
        .map_err(|e| AppError::ApiError(e.to_string()))?;

    let feeds: Vec<FeedInfo> = response
        .data
        .feeds
        .iter()
        .map(|f| FeedInfo {
            uri: f.uri.to_string(),
            cid: f.cid.as_ref().to_string(),
            did: f.did.to_string(),
            creator_did: f.creator.did.to_string(),
            creator_handle: f.creator.handle.to_string(),
            creator_display_name: f.creator.display_name.clone(),
            creator_avatar: f.creator.avatar.clone(),
            display_name: f.display_name.clone(),
            description: f.description.clone(),
            avatar: f.avatar.clone(),
            like_count: f.like_count.unwrap_or(0) as u32,
            is_saved: f.viewer.as_ref().and_then(|v| v.like.as_ref()).is_some(),
        })
        .collect();

    Ok(SuggestedFeedsResponse {
        feeds,
        cursor: response.data.cursor,
    })
}

#[derive(Deserialize)]
pub struct GetFeedRequest {
    pub feed_uri: String,
    pub limit: Option<u8>,
    pub cursor: Option<String>,
}

#[derive(Serialize)]
pub struct FeedPostsResponse {
    pub posts: Vec<super::timeline::TimelinePost>,
    pub cursor: Option<String>,
}

/// Get posts from a specific feed
#[tauri::command]
pub async fn get_feed(
    app: AppHandle,
    agent_state: State<'_, AgentState>,
    request: GetFeedRequest,
) -> Result<FeedPostsResponse, AppError> {
    println!("DEBUG: get_feed called with uri: {}", request.feed_uri);

    let guard = agent_state.lock().await;
    let agent = guard.as_ref().ok_or(AppError::SessionNotFound)?;

    let limit_val = request.limit.unwrap_or(50).max(1).min(100);
    let limit = bsky_sdk::api::types::LimitedNonZeroU8::<100>::try_from(limit_val).ok();

    let response = agent
        .api
        .app
        .bsky
        .feed
        .get_feed(
            bsky_sdk::api::app::bsky::feed::get_feed::ParametersData {
                feed: request
                    .feed_uri
                    .parse()
                    .map_err(|_| AppError::ApiError("Invalid feed URI".into()))?,
                cursor: request.cursor,
                limit,
            }
            .into(),
        )
        .await
        .map_err(|e| {
            println!("DEBUG: get_feed API error: {}", e);
            AppError::ApiError(e.to_string())
        })?;

    println!(
        "DEBUG: get_feed success, items: {}",
        response.data.feed.len()
    );

    let mut posts: Vec<super::timeline::TimelinePost> = Vec::new();
    for feed_view in &response.data.feed {
        let post = &feed_view.post;

        let text = if let Ok(json) = serde_json::to_value(&post.record) {
            json.get("text")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string()
        } else {
            String::new()
        };

        let created_at = if let Ok(json) = serde_json::to_value(&post.record) {
            json.get("createdAt")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string()
        } else {
            String::new()
        };

        let embed = media::process_post_embed(post, &app).await?;
        let (is_repost, reposted_by_handle, reposted_by_display_name) =
            super::timeline::extract_repost_context(feed_view);

        posts.push(super::timeline::TimelinePost {
            uri: post.uri.to_string(),
            cid: post.cid.as_ref().to_string(),
            author_did: post.author.did.to_string(),
            author_handle: post.author.handle.to_string(),
            author_display_name: post.author.display_name.clone(),
            author_avatar: post.author.avatar.clone(),
            is_repost,
            reposted_by_handle,
            reposted_by_display_name,
            text,
            created_at,
            reply_count: post.reply_count.unwrap_or(0) as u32,
            repost_count: post.repost_count.unwrap_or(0) as u32,
            like_count: post.like_count.unwrap_or(0) as u32,
            is_liked: post.viewer.as_ref().and_then(|v| v.like.as_ref()).is_some(),
            is_reposted: post
                .viewer
                .as_ref()
                .and_then(|v| v.repost.as_ref())
                .is_some(),
            viewer_like: post
                .viewer
                .as_ref()
                .and_then(|v| v.like.as_ref())
                .map(|u| u.to_string()),
            viewer_repost: post
                .viewer
                .as_ref()
                .and_then(|v| v.repost.as_ref())
                .map(|u| u.to_string()),
            embed: embed.and_then(|value| serde_json::to_value(value).ok()),
        });
    }

    Ok(FeedPostsResponse {
        posts,
        cursor: response.data.cursor,
    })
}
